internal/api/security.go

```diff
@@ -0,0 +1,276 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+//
+
+package api
+
+import (
+	"net/http"
+
+	"github.com/gorilla/mux"
+)
+
+// initSecurity registers security endpoints on the given router.
+func initSecurity(apiRouter *mux.Router, context *Context) {
+	addContext := func(handler contextHandlerFunc) *contextHandler {
+		return newContextHandler(context, handler)
+	}
+
+	securityRouter := apiRouter.PathPrefix("/security").Subrouter()
+
+	securityClusterRouter := securityRouter.PathPrefix("/cluster/{cluster:[A-Za-z0-9]{26}}").Subrouter()
+	securityClusterRouter.Handle("/api/lock", addContext(handleClusterLockAPI)).Methods("POST")
+	securityClusterRouter.Handle("/api/unlock", addContext(handleClusterUnockAPI)).Methods("POST")
+
+	securityInstallationRouter := securityRouter.PathPrefix("/installation/{installation:[A-Za-z0-9]{26}}").Subrouter()
+	securityInstallationRouter.Handle("/api/lock", addContext(handleInstallationLockAPI)).Methods("POST")
+	securityInstallationRouter.Handle("/api/unlock", addContext(handleInstallationUnockAPI)).Methods("POST")
+
+	securityClusterInstallationRouter := securityRouter.PathPrefix("/cluster_installation/{cluster_installation:[A-Za-z0-9]{26}}").Subrouter()
+	securityClusterInstallationRouter.Handle("/api/lock", addContext(handleClusterInstallationLockAPI)).Methods("POST")
+	securityClusterInstallationRouter.Handle("/api/unlock", addContext(handleClusterInstallationUnockAPI)).Methods("POST")
+
+	securityGroupRouter := securityRouter.PathPrefix("/group/{group:[A-Za-z0-9]{26}}").Subrouter()
+	securityGroupRouter.Handle("/api/lock", addContext(handleGroupLockAPI)).Methods("POST")
+	securityGroupRouter.Handle("/api/unlock", addContext(handleGroupUnockAPI)).Methods("POST")
+}
+
+// handleClusterLockAPI responds to POST /api/cluster/{cluster}/api/lock,
+// locking API changes for this cluster.
+func handleClusterLockAPI(c *Context, w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	clusterID := vars["cluster"]
+	c.Logger = c.Logger.WithField("cluster", clusterID)
+
+	cluster, err := c.Store.GetCluster(clusterID)
+	if err != nil {
+		c.Logger.WithError(err).Error("failed to query cluster")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	if cluster == nil {
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
+
+	if !cluster.APISecurityLock {
+		err = c.Store.LockClusterAPI(cluster.ID)
+		if err != nil {
+			c.Logger.WithError(err).Error("failed to lock cluster API")
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	}
+
+	w.WriteHeader(http.StatusOK)
+}
+
+// handleClusterUnockAPI responds to POST /api/cluster/{cluster}/api/unlock,
+// unlocking API changes for this cluster.
+func handleClusterUnockAPI(c *Context, w http.ResponseWriter, r *http.Request) {
```

Comment:
UnockAPI -> UnlockAPI

Reply:
Nice catch! Who knows what an Unock even is. ;)

Reply:
it's what happens when you arrive at a friend's door